src/openssl/ssl/hostname_validation.cr

```diff
@@ -0,0 +1,128 @@
+require "socket"
+require "openssl"
+
+# :nodoc:
+module OpenSSL::SSL::HostnameValidation
+  enum Result
+    Error
+    MalformedCertificate
+    MatchFound
+    MatchNotFound
+    NoSANPresent
+  end
+
+  # Matches hostname against Subject Alternate Name (SAN) or Common Name (CN)
+  # entries of certificate if no SAN entries could be found in the certificate.
+  def self.validate_hostname(hostname : String, server_cert : LibCrypto::X509)
+    return Result::Error if server_cert.null?
+    result = matches_subject_alternative_name(hostname, server_cert)
+    result = matches_common_name(hostname, server_cert) if result.no_san_present?
```

Comment:
Are we sure that we need to require the hostname in the SANs if they're present? So far my understand was that while it's recommended to repeat the CN in the SANs, it's not required.